### PR TITLE
[MLv2] Fix some edge cases and add more unit tests for zoom in drills

### PR DESCRIPTION
**Pull request by bshepherdson**
More progress towards #36253.

### CODE REVIEW

**Comment by bshepherdson** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/38275?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #38210** <a href="https://app.graphite.dev/github/pr/metabase/metabase/38210?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #38275** <a href="https://app.graphite.dev/github/pr/metabase/metabase/38275?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/38275?utm_source=stack-comment).

File: test/metabase/lib/drill_thru/zoom_in_bins_test.cljc [hunk 1]
```diff
@@ -117,3 +129,100 @@
                                           [:field {:binning {:strategy :default}} (meta/id :orders :quantity)]
                                           32.5]]}]}
                 (lib/drill-thru query -1 drill)))))))
+
+;; TODO: Add a test for clicking on pivot column cells (and headers?) - but that's broken on master. See #38265.
+(deftest ^:parallel cell-click-filters-and-updates-only-one-dimension-test
+  (testing "when zooming in on one dimension, other breakouts are dropped"
```
Comment (metamben) [inline test/metabase/lib/drill_thru/zoom_in_bins_test.cljc:135]: Where can I see that the other breakouts get dropped?

File: test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc [hunk 2]
```diff
@@ -4,13 +4,25 @@
...
+(deftest ^:parallel zoom-in-timeseries-available-test
+  (testing "zoom-in for bins is available for cells, pivots and legends on numeric columns which have binning set"
+    (doseq [[test-case context {:keys [click]}] (canned/canned-clicks)]
+      (if (and (#{:cell :pivot :legend} click)
+               (not-empty (for [dim (:dimensions context)
+                                :when (and (isa? (:effective-type (:column dim)) :type/DateTime)
+                                           (lib/temporal-bucket (:column dim)))]
+                            dim)))
+        (is (canned/returned test-case context :drill-thru/zoom-in.timeseries))
+        (is (not (canned/returned test-case context :drill-thru/zoom-in.timeseries)))))))
```
Comment (metamben) [inline test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc:24]: Like the other test.

File: test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc [hunk 3]
```diff
@@ -164,3 +176,37 @@
                                               [:field {:temporal-unit :year} (meta/id :orders :created-at)]
                                               "2022-12-01T00:00:00+02:00"]]}]}
                     (lib/drill-thru query -1 drill)))))))))
+
+(def ^:private unit-pairs
+  (let [units @#'lib.drill-thru.zoom-in-timeseries/valid-current-units]
+    (map vector units (rest units))))
```
Comment (metamben) [inline test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc:182]: I think the standard way to write this is
```suggestion
    (partition 2 1 units)))
```
It doesn't produce vectors, but that seems to be good enough here.

**Review by metamben — APPROVED** [review]
LGTM, there's just one test that I didn't quite get.

File: test/metabase/lib/drill_thru/zoom_in_bins_test.cljc [hunk 4]
```diff
@@ -4,11 +4,23 @@
...
 #?(:cljs (comment metabase.test-runner.assert-exprs.approximately-equal/keep-me))
 
+(deftest ^:parallel zoom-in-bins-available-test
+  (testing "zoom-in for bins is available for cells, pivots and legends on numeric columns which have binning set"
+    (doseq [[test-case context {:keys [click]}] (canned/canned-clicks)]
+      (if (and (#{:cell :pivot :legend} click)
+               (not-empty (for [dim (:dimensions context)
+                                :when (and (isa? (:effective-type (:column dim)) :type/Number)
+                                           (lib/binning (:column dim)))]
+                            dim)))
```
[the remark this answers is not in the document]
Comment (bshepherdson) [reply test/metabase/lib/drill_thru/zoom_in_bins_test.cljc:20]: I did it this way on purpose - you get nicer output when it fails that indicates the query and context that were used.

File: test/metabase/lib/drill_thru/zoom_in_bins_test.cljc (hunk 1 above)
Comment (bshepherdson) [reply test/metabase/lib/drill_thru/zoom_in_bins_test.cljc:135]: This is badly phrased. It's more like "existing breakouts are dropped, filters are added, and new breakouts are added", so there are still breakouts in the final query.

I've updated the wording, and improved the test to show that an existing `:num-bins 10` gets turned into a `:strategy :default`.

File: test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc (hunk 3 above)
Comment (bshepherdson) [reply test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc:182]: That works for me!

**Comment by bshepherdson** [comment]
### Merge activity

* **Feb 2, 10:26 AM EST**: @bshepherdson merged this pull request with [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/38275).

**Pull request merged**